### ISSUE TITLE
Move together features out of scenery/axon/sun/etc.

### ISSUE THREAD

**Issue by samreid** [phetsims/together#72]
@jonathanolson suggested moving together features out of scenery/axon/sun/etc.

For instance, in Property, we could have:
this.trigger0( 'startedNotifyingObservers' );
...
this.trigger0( 'endedNotifyingObservers' );

And instead of component registration being in base classes such as RadioButton, we would have 
```js
together.addComponent( 'beakerScreen.runTrialButton', runTrialButton );
```

He also suggested making components composite, such as ComboBox would just appear as a single UI component, instead of one-per-selection-button.

**Comment by samreid** [comment]
About performance:

arch && arch.start() would get stripped out

But:
this.trigger0('startedProcessingDragStartedListeners');
this.trigger0('endedProcessingDragStartedListeners');

Would not be able to be stripped out.  So we would want to make sure that it is OK to add 2+ new axon triggers for *every* drag event.

**Comment by samreid** [comment]
@jbphet said we could still use a global directive for stripping out those values, like "dc" for data collection, if we discover that performance is an issues.

dc && this.trigger0(...)

**Comment by samreid** [comment]
We could even move the together&&together.addComponent() out of the sim code--perhaps make it part of a sim object registry that together can listen to.  Then there will be no together references in the sim code or the sim common libraries.

**Comment by samreid** [comment]
I think this is a good suggestion and am planning to move forward with it.

**Comment by samreid** [comment]
One more step in isolating/modularizing the together code could be: create a registry for each sim screen where objects can "register" themselves with that screen.  This would be part of joist.  Then together could observe that registry to add together-specific listeners. handling, support, etc.

**Comment by samreid** [comment]
I said screen above since I thought we might be able to auto-associate elements with their screen--however, I'm not sure how this would work and perhaps association at the sim level (instead of the screen) level would be more practical.

**Comment by samreid** [comment]
I sent out this email regarding this and #73:

> Based in part on a discussion I had yesterday with John Blanco and Jon Olson, I'll be making significant changes to the together repository as well as how beers-law-lab, sun, scenery-phet, axon are instrumented for together interoperability.  Jon Olson described a way to generalize the implementation to keep together modularized and prevent together logic from leaking into our common libraries. It looks good but it is a large refactor and I have been on a dirty working copy for 24+ hours now.  I'll let you know when it is ready to commit, and it will need some attention for review.

> Best Regards,
> Sam

**Comment by samreid** [comment]
Since I don't want to tangle with branches from 6 repos, I'm just trying to get my working copy changes to a point where I can commit them all to master with low enough risk that we can continue in master.

**Comment by samreid** [comment]
I committed the largest part of this refactor above.  Common libraries and beers-law-lab are passing JSHint, and I tested Reactants Products and Leftovers and Beers Law Lab and didn't see any obvious trouble.  The changes above should be reviewed and discussed at some point--we should discuss who should review and how much time to spend on it at an upcoming developer meeting.

**Comment by samreid** [comment]
In the above commit, faradays-law is back to full feature set for mirroring + data stream.

**Comment by pixelzoom** [comment]
#72 and #73 sound like big changes in approach, and those changes aren't clearly described in the issues. And since I'm on vacation, I don't have the tools necessary to dig around in code to formulate a picture of what's changed. So I'll trust that these changes are a really good thing and don't paint us into any corners (although statements like "ComboBox would just appear as a single UI component" seem a little limiting). If someone has time on Thursday 4/16, I'd appreciate a summary to catch me up.

**Comment by samreid** [comment]
Well, in #79 we decided that scenery-phet and other phet-specific libraries should register components.  Axon, Sun, Scenery and other "general" libraries will instead provide an alternative way to do this.

At today's meeting we discussed the alternative approach of making its constituents available for clients to use.  But this seems unsatisfying since it breaks encapsulation and differs from the approach elsewhere.  Perhaps we can make together integration possible in sun/axon/etc possible but optional?  I was considering a repository-specific registry that could be enabled so that all sun components would register themselves, and together could "hook in" to that.
